src/Avalonia.Base/Input/GamepadManager.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Avalonia.Interactivity;
+using Avalonia.Metadata;
+using Avalonia.Reactive;
+using Avalonia.Threading;
+
+namespace Avalonia.Input
+{
+    [NotClientImplementable()]
+    public interface IGamepadManager
+    {
+        /// <summary>
+        /// Obtains a stream of gamepad events. 
+        /// </summary>
+        IObservable<GamepadEventArgs> GamepadStream { get; }
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <returns></returns>
+        IReadOnlyList<GamepadState> GetSnapshot();
+    }
+
+    public abstract class GamepadManager : IGamepadManager
+    {
+        /// <summary>
+        /// The static entry-point to interacting with Gamepads. This is null if the current platform does not have a Gamepad implementation. 
+        /// </summary>
+        public static GamepadManager? Instance { get; protected set; }
+
+        public static readonly RoutedEvent GamepadStateChanged;
+        static GamepadManager()
+        {
+            GamepadStateChanged = RoutedEvent.Register<GamepadManager, GamepadEventArgs>(nameof(GamepadStateChanged), RoutingStrategies.Tunnel | RoutingStrategies.Bubble);
+        }
+
+        private readonly LightweightSubject<GamepadEventArgs> _gamepadStream = new();
+        protected readonly List<GamepadState> _currentState = [];
+
+        public void PushGamepadEvent(GamepadEventArgs args)
+        {
+            if (_currentState.Count <= args.Device)
+            {
+                _currentState.Add(args.State);
+            }
+            else
+            {
+                _currentState[args.Device] = args.State;
+            }
+
+            // question - if I use Post instead of Invoke will events end up out of order?
+            Dispatcher.UIThread.Post(() =>
+            {
+                _gamepadStream.OnNext(args);
+            });
+        }
+
+        public IObservable<GamepadEventArgs> GamepadStream => _gamepadStream;
```

Comment:
So, is it supposed to be routed event API or observable API?
We will need to discuss this API on the API review later. But I am just asking early questions.

Reply:
> So, is it supposed to be routed event API or observable API? We will need to discuss this API on the API review later. But I am just asking early questions.

Yeah, I am being pulled in both directions - the Observable API is great for streaming these events, but if we want them to route on the UI layer, they would need to be routable events. 

I am considering splitting these two up, with an IObservable<GamepadEventArgs> that doesn't get routed, and if not "handled" at this level, producing routed-events from these - in line with what TomEdwards was saying above - turning these "raw" gamepad events into Semantic Avalonia events that can work with a UI.